src/lib/pubkey/classic_mceliece/cmce_encaps.cpp

```diff
@@ -0,0 +1,126 @@
+            return std::nullopt;
+         }
+      }
+   }
+
+   secure_vector<uint8_t> a_value_byte(params.t());
+   secure_vector<uint8_t> e_bytes(ceil_tobytes(params.n()));
+
+   // Step 5: Set all bits of e at the positions of a_values
+   // Prepare the associated byte in e_bytes that is represented by each bit index in a_values
+   // if we e is represented as a byte vector
+   for(size_t j = 0; j < a_values.size(); ++j) {
+      a_value_byte[j] = 1 << (a_values[j] % 8);
+   }
+
+   for(size_t i = 0; i < params.n() / 8; ++i) {
+      for(size_t j = 0; j < a_values.size(); ++j) {
+         // If the current byte is the one that is represented by the current bit index in a_values
+         // then set the bit in e_bytes (in-byte position prepared above)
+         auto mask = CT::Mask<uint16_t>::is_equal(static_cast<uint16_t>(i), static_cast<uint16_t>(a_values[j] >> 3));
+         e_bytes[i] |= mask.if_set_return(a_value_byte[j]);
+      }
+   }
+
+   return secure_bitvector(e_bytes, params.n());
+}
+
+void Classic_McEliece_Encryptor::raw_kem_encrypt(std::span<uint8_t> out_encapsulated_key,
+                                                 std::span<uint8_t> out_shared_key,
+                                                 RandomNumberGenerator& rng) {
+   BOTAN_ASSERT(out_encapsulated_key.size() == m_key->params().ciphertext_size(),
+                "Correct encapsulated key output length");
+   BOTAN_ASSERT(out_shared_key.size() == m_key->params().hash_out_bytes(), "Correct shared key output length");
+
+   auto& params = m_key->params();
+
+   // Call fixed_weight until it is successful to
+   // create a random error vector e of weight tau
+   secure_bitvector e = [&] {
+      // Emergency abort in case unexpected logical error to prevent endless loops
+      //   Success probability: >24% per attempt (25% that elements are distinct * 96% enough elements are in range)
+      //   => 203 attempts for 2^(-80) fail probability
+      const size_t MAX_ATTEMPTS = 203;
+      for(size_t attempt = 0; attempt < MAX_ATTEMPTS; ++attempt) {
+         if(auto maybe_e = fixed_weight_vector_gen(params, rng.random_vec((params.sigma1() / 8) * params.tau()))) {
+            return maybe_e.value();
+         }
+      }
+      throw Internal_Error("Cannot created fixed weight vector.");
+   }();
+
+   auto big_c = encode(params, e, m_key->matrix()).to_bytes();
+   auto hash_func = params.hash_func();
+
+   if(params.is_pc()) {
+      hash_func->update(2);
+      hash_func->update(e.to_bytes());
+      auto big_c_1 = hash_func->final_stdvec();
+      big_c = Botan::concat(big_c, big_c_1);
+      hash_func->clear();
+   }
+
+   hash_func->update(1);
+   hash_func->update(e.to_bytes());
+   hash_func->update(big_c);
+
+   auto big_k = hash_func->final<secure_vector<uint8_t>>();
+   hash_func->clear();
+
+   std::copy(big_c.begin(), big_c.end(), out_encapsulated_key.begin());
+   std::copy(big_k.begin(), big_k.end(), out_shared_key.begin());
```

Comment:
Why not write straight into the output buffers? Also note that we can rely on `hash_func.final...()` calling `clear()`, so we don't need to do that explicitly.

```suggestion
   auto hash_func = params.hash_func();
   auto big_c = [&] {
      BufferStuffer ss_out(out_encapsulated_key);
      const auto encoded_matrix = encode(params, e, m_key->matrix());
      encoded_matrix.to_bytes(ss_out.next(ceil_tobytes(encoded_matrix.size())));

      if(params.is_pc()) {
         hash_func->update(0x02);
         hash_func->update(e.to_bytes());
         hash_func->final(ss_out.next(hash_func->output_length()));
      }

      BOTAN_ASSERT_NOMSG(ss_out.full());
      return out_encapsulated_key;
   }();

   hash_func->update(0x01);
   hash_func->update(e.to_bytes());
   hash_func->update(big_c);
   hash_func->final(out_shared_key);
```